### PR TITLE
feat: tagging exported content records for transmission

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
https://2u-internal.atlassian.net/browse/ENT-6339

exporters - always tag content records for transmission
transmitters - always removes tag on content records after transmitting

cornerstone 
                      - now included with all other channels for regularly scheduled transmissions
                      - now has a dummy transmitter transmit method
                      - now has a `transmit for web` transmitter method that functions like the normal transmitter

during scheduled transmissions, cornerstone will default to the dummy transmitter method that will result in a noop. During pulls, it will use the regular transmitter transmit method

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/cornerstone/exporters/content_metadata.py [hunk 1]
```diff
@@ -53,13 +53,7 @@ def export_for_web_polling(self, max_payload_count=MAX_PAYLOAD_COUNT):
         """
         Return the exported and transformed content metadata as a dictionary for CSDO web pull.
         """
-        return self.export(max_payload_count=max_payload_count)
-
-    def export_force_all_catalogs(self):
```
Comment (alex-sheehan-edx) [inline integrated_channels/cornerstone/exporters/content_metadata.py:58]: this isn't used by anything anymore

Comment (johnnagro) [inline integrated_channels/cornerstone/exporters/content_metadata.py:58]: note - we dont use this anymore which is why we're removing it

File: integrated_channels/cornerstone/exporters/content_metadata.py [hunk 2]
```diff
@@ -53,13 +53,7 @@ def export_for_web_polling(self, max_payload_count=MAX_PAYLOAD_COUNT):
...
-        """
-        Return the exported and transformed content metadata as a dictionary regardless if there is an update needed.
-        """
-        return self.export(force_retrieve_all_catalogs=True)
+        return self.export(max_payload_count=max_payload_count, enable_cache=True)
```
Comment (johnnagro) [inline integrated_channels/cornerstone/exporters/content_metadata.py:56]: i'd pipe `enable_cache` in as a kwarg for `export_for_web_polling`

File: integrated_channels/cornerstone/transmitters/content_metadata.py [hunk 3]
```diff
@@ -16,12 +16,27 @@ def __init__(self, enterprise_configuration, client=CornerstoneAPIClient):
             client=client
         )
 
-    def _transmit_action(self, content_metadata_item_map, client_method, action_name):
+    def transmit(self, create_payload, update_payload, delete_payload, **kwargs):
```
Comment (johnnagro) [inline integrated_channels/cornerstone/transmitters/content_metadata.py:19]: i dont personally think we should handle the cache via a kwarg here. i'll huddle and suggest an alternative

File: integrated_channels/integrated_channel/models.py [hunk 4]
```diff
@@ -518,6 +525,37 @@ def incomplete_delete_transmissions(
         )
         return ContentMetadataItemTransmission.objects.filter(in_db_but_failed_to_send_query)
 
+    def mark_transmission(self, mark_for):
```
Comment (johnnagro) [inline integrated_channels/integrated_channel/models.py:528]: `mark_transmission` -> `_mark_transmission` if its an internal method, given the existence of things like `mark_for_create`

**Review by johnnagro — CHANGES_REQUESTED** [review]
lets huddle

File: integrated_channels/cornerstone/transmitters/content_metadata.py (hunk 3 above)
Comment (johnnagro) [reply integrated_channels/cornerstone/transmitters/content_metadata.py:19]: if we use a kwarg here to skip clearing of these marked records, lets make sure it also causes us to skip updating their timestamps, api reponses, and error messages

File: integrated_channels/integrated_channel/transmitters/content_metadata.py [hunk 5]
```diff
@@ -87,14 +87,15 @@ def transmit(self, create_payload, update_payload, delete_payload, **kwargs):
         Transmit content metadata items to the integrated channel. Save or update content metadata records according to
         the type of transmission.
         """
+        remove_mark_for = kwargs.get('remove_mark_for', True)
         self._log_info_for_each_item_map(delete_payload, 'transmitting delete')
-        delete_payload_results = self._transmit_delete(delete_payload)
+        delete_payload_results = self._transmit_delete(delete_payload, remove_mark_for=remove_mark_for)
```
Comment (johnnagro) [inline integrated_channels/integrated_channel/transmitters/content_metadata.py:92]: perhaps pass `**kwargs` instead of a specific kwarg?

**Comment by johnnagro** [comment]
@alex-sheehan-edx and I have had a few huddles to discuss this PR as well as thought-experiment possible alternative approaches. I think we've landed on something good and he's going to send that draft out for review shortly.

File: integrated_channels/cornerstone/transmitters/content_metadata.py [hunk 6]
```diff
@@ -16,6 +16,18 @@ def __init__(self, enterprise_configuration, client=CornerstoneAPIClient):
             client=client
         )
 
+    def transmit(self, create_payload, update_payload, delete_payload):
```
Comment (alex-sheehan-edx) [inline integrated_channels/cornerstone/transmitters/content_metadata.py:19]: cornerstone transmitter.transmit -> does nothing
cornerstone transmitter.trasnmit_for_web -> does what the normal transmit does

File: integrated_channels/integrated_channel/exporters/content_metadata.py [hunk 7]
```diff
@@ -209,39 +211,43 @@ def _check_matched_content_to_create(
                 integrated_channel_code=self.enterprise_configuration.channel_code(),
                 content_id=content_id,
             ).first()
-            incomplete_transmission = ContentMetadataItemTransmission.incomplete_create_transmissions(
```
Comment (alex-sheehan-edx) [inline integrated_channels/integrated_channel/exporters/content_metadata.py:212]: This lookup doesn't have to happen unless we've verified that the previous lookup (`past_deleted_transmission`) does not exist, so I moved it to after the `else`

File: integrated_channels/integrated_channel/management/commands/__init__.py [hunk 8]
```diff
@@ -45,6 +45,7 @@
     for integrated_channel_class in (
         BlackboardEnterpriseCustomerConfiguration,
         CanvasEnterpriseCustomerConfiguration,
+        CornerstoneEnterpriseCustomerConfiguration,
```
Comment (alex-sheehan-edx) [inline integrated_channels/integrated_channel/management/commands/__init__.py:48]: Cornerstone would now be included in the regularly scheduled transmit content task

File: integrated_channels/cornerstone/exporters/content_metadata.py [hunk 9]
```diff
@@ -49,17 +49,18 @@ class CornerstoneContentMetadataExporter(ContentMetadataExporter):
...
-    def export_force_all_catalogs(self):
-        """
-        Return the exported and transformed content metadata as a dictionary regardless if there is an update needed.
-        """
-        return self.export(force_retrieve_all_catalogs=True)
+        ContentMetadataItemTransmission = apps.get_model(
+            'integrated_channel',
+            'ContentMetadataItemTransmission'
+        )
+        create_payload = ContentMetadataItemTransmission.objects.filter(marked_for='create')
```
Comment (johnnagro) [inline integrated_channels/cornerstone/exporters/content_metadata.py:60]: we also need to filter by `enterprise_customer` and `plugin_config_id`

File: integrated_channels/integrated_channel/management/commands/__init__.py (hunk 8 above)
Comment (johnnagro) [reply integrated_channels/integrated_channel/management/commands/__init__.py:48]: you need to update the comment on line 41/42

File: integrated_channels/cornerstone/transmitters/content_metadata.py (hunk 3 above)
Comment (johnnagro) [reply integrated_channels/cornerstone/transmitters/content_metadata.py:19]: you can probably drop the `**kwargs`

File: integrated_channels/cornerstone/transmitters/content_metadata.py [hunk 10]
```diff
@@ -16,6 +16,18 @@ def __init__(self, enterprise_configuration, client=CornerstoneAPIClient):
             client=client
         )
 
+    def transmit(self, create_payload, update_payload, delete_payload, **kwargs):
+        """
+        Placholder
+        """
+        pass
+
+    def transmit_for_web(self, create_payload, update_payload, delete_payload, **kwargs):
```
Comment (johnnagro) [inline integrated_channels/cornerstone/transmitters/content_metadata.py:25]: i think you can drop the `**kwargs`

File: integrated_channels/cornerstone/transmitters/content_metadata.py [hunk 11]
```diff
@@ -16,6 +16,18 @@ def __init__(self, enterprise_configuration, client=CornerstoneAPIClient):
...
+        """
+        Placeholder
+        """
+        return super().transmit(create_payload, update_payload, delete_payload, **kwargs)
```
Comment (johnnagro) [inline integrated_channels/cornerstone/transmitters/content_metadata.py:29]: drop dem `**kwargs`

File: integrated_channels/cornerstone/transmitters/content_metadata.py [hunk 12]
```diff
@@ -16,6 +16,18 @@ def __init__(self, enterprise_configuration, client=CornerstoneAPIClient):
             client=client
         )
 
+    def transmit(self, create_payload, update_payload, delete_payload, **kwargs):
+        """
+        Placholder
+        """
+        pass
```
Comment (johnnagro) [inline integrated_channels/cornerstone/transmitters/content_metadata.py:23]: lets also log that this has run and we're deliberately doing nothing, i think you can use the `_log_info` here to properly pass all the variables for splunk

**Review by johnnagro — CHANGES_REQUESTED** [review]
looking really good! one critical thing needs to change  - your filtering on the cached export needs to be customer, plugin-specific. otherwise some nits.

File: integrated_channels/cornerstone/exporters/content_metadata.py (hunk 9 above)
Comment (alex-sheehan-edx) [reply integrated_channels/cornerstone/exporters/content_metadata.py:60]: agreed

File: integrated_channels/cornerstone/views.py [hunk 13]
```diff
@@ -127,8 +127,14 @@ def get(self, request, *args, **kwargs):
 
         get_count = int(request.GET.get('count', exporter.MAX_PAYLOAD_COUNT))
         content_item_count = min(get_count, exporter.MAX_PAYLOAD_COUNT)
-        created, updated, deleted = transmitter.transmit(
-            *exporter.export_for_web_polling(max_payload_count=content_item_count)
+        # Remove content records `marked for transmission` flags
+        created, updated, deleted = exporter.export_for_web_polling(max_payload_count=content_item_count)
+
+        create_payload = {record.content_id: record for record in created}
```
Comment (johnnagro) [inline integrated_channels/cornerstone/views.py:133]: ah. this should be inside the `export_for_web_polling()` method. the export signature should be consistent - including the structure of its return values.

Comment (alex-sheehan-edx) [reply integrated_channels/cornerstone/views.py]: done

**Review by johnnagro — APPROVED** [review]
⛵

**Pull request merged**